### PR TITLE
[docs] Improve Tune + Datasets documentation

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Improve the Tune/Datasets docs to cover the configuration options and have code examples of using Datasets inside and outside placement groups.

Split out from https://github.com/ray-project/ray/pull/25241/files

### CODE REVIEW

File: doc/source/data/doc_code/key_concepts.py [hunk 1]
```diff
@@ -1,32 +1,57 @@
...
+
+# This Dataset workload will use spare cluster resources for execution.
+def objective(*args):
+    ray.data.range(10).show()
+
+# Create a cluster with 4 CPU slots available.
+ray.init(num_cpus=4)
+
+# This runs, since Tune schedules one trial on 1 CPU, leaving 3 spare CPUs in the
+# cluster for Dataset execution. However, deadlock can occur if you set num_samples=4,
```
Comment (jianoaix) [inline doc/source/data/doc_code/key_concepts.py:16]: Can the same problem occur if Tune and Dataset compete for resources within the PG like they compete within cluster?

**Review by jianoaix — APPROVED** [review]

File: doc/source/data/doc_code/key_concepts.py (hunk 1 above)
Comment (ericl) [reply doc/source/data/doc_code/key_concepts.py:16]: Technically no, assuming your PG is properly configured. If you configure it wrong then yes :)

**Comment by clarkzinzow** [comment]
@ericl This is a great docs addition, thank you for adding this! 🙌

**Comment by clarkzinzow** [comment]
@ericl Did this break lint?

**Comment by ericl** [comment]
Seems @krfricke fixed it (thanks!) I didn't see the LINT error since there was another LINT failure in master at the tune :(

**Pull request merged**